src/OpenTelemetry/Metrics/DoubleObserverMetricSdk.cs

```diff
@@ -35,8 +35,12 @@ public DoubleObserverMetricSdk(string name, Action<DoubleObserverMetric> callbac
         public override void Observe(double value, LabelSet labelset)
         {
             // TODO cleanup of handle/aggregator.   Issue #530
-            var boundInstrument =
-                this.observerHandles.GetOrAdd(labelset, new DoubleObserverMetricHandleSdk());
+
+            DoubleObserverMetricHandleSdk boundInstrument;
+            if (!this.observerHandles.TryGetValue(labelset, out boundInstrument))
+            {
+                boundInstrument = this.observerHandles.GetOrAdd(labelset, new DoubleObserverMetricHandleSdk());
```

Comment:
These other ones that don't need a closure you could use a static delegate:
```csharp
        private static readonly Func<LabelSet, DoubleObserverMetricHandleSdk> CreateMetricFunc = (labelSet) => new DoubleObserverMetricHandleSdk();

        public override void Observe(double value, LabelSet labelset)
        {
            // TODO cleanup of handle/aggregator.   Issue #530
            var boundInstrument =
                this.observerHandles.GetOrAdd(labelset, CreateMetricFunc);

            boundInstrument.Observe(value);
        }
```

Reply:
I'm not sure this is necessary as we already called TryGetValue prior.  Thus, the expectation is that we will need to alloc the new item.  Thus, converting to a static delegate is just adding 1 extra layer.

If your concern is about atomicity, then, both these are equivalent.  In the case of a race condition, the GetOrAdd(Func<>) form does not guarantee atomicity either.  Thus, the functionally is no different than current code.  In both cases, we will wind up "leaking" the new object regardless.

IMHO, In this code base, I would prefer using the GetOrAdd(Func<>) syntax rather than break into TryGetValue/TryGetOrAdd pattern.  Simply because the code is more readable and understandable.  Plus, the compiler has the ability to optimize to "static" delegates as it sees fit.  Also, going forward in newer .NET, we have the GetOrAdd syntax with passing in TArg, which addresses the performance with closure.

Reply:
> I'm not sure this is necessary as we already called TryGetValue prior. Thus, the expectation is that we will need to alloc the new item. Thus, converting to a static delegate is just adding 1 extra layer.

Don't call TryGetValue for the case where a static delegate will work, just call TryGetValue passing the delegate ref (most places). Only call TryGetValue/TryAdd when you need the closure (the one place).

> IMHO, In this code base, I would prefer using the GetOrAdd(Func<>) syntax rather than break into TryGetValue/TryGetOrAdd pattern. 

Let's go with the allocation-free patterns.